src/Client/Updater.php

```diff
@@ -324,4 +307,103 @@ protected function verifySingleSignature(string $bytes, array $signatureMeta)
+            $rootsDownloaded++;
+            if ($rootsDownloaded > static::MAX_ROOT_DOWNLOADS) {
+                throw new \Exception("The maximum number root files have already been dowloaded:" . static::MAX_ROOT_DOWNLOADS);
+            }
+            $nextRoot = RootMetadata::createFromJson($nextRootContents);
+            // SPEC: 1.3
+            $this->checkSignatures($nextRoot);
+            // Update Role and Key databases to use the new root information.
+            $this->roleDB = RoleDB::createRoleDBFromRootMetadata($nextRoot);
+            $this->keyDB = KeyDB::createKeyDBFromRootMetadata($nextRoot);
+            $this->checkSignatures($nextRoot);
+            // SPEC: 1.4
+            $this->checkRollbackAttack($rootData, $nextRoot, $nextVersion);
+            $rootData = $nextRoot;
+            // SPEC: 1.5 - Needs no action.
+            // Note that the expiration of the new (intermediate) root metadata
+            // file does not matter yet, because we will check for it in step
+            // 1.8.
+
+            // SPEC: 1.6
+            $this->durableStorage['root.json'] = $nextRootContents;
+            $nextVersion = $rootData->getVersion() + 1;
+            // SPEC: 1.7 Repeat the above steps.
+        }
+        // SPEC: 1.8
+        $this->checkFreezeAttack($rootData, $this->getCurrentTime());
+
+        // SPEC 1.9: Delete the trusted timestamp and snapshot files if either
+        // file has rooted keys.
+        if ($rootsDownloaded &&
+           ($this->hasRotatedKeys($originalRootData, $rootData, 'timestamp')
+           || $this->hasRotatedKeys($originalRootData, $rootData, 'snapshot'))) {
+            unset($this->durableStorage['timestamp.json'], $this->durableStorage['snapshot.json']);
+        }
+    }
+
+    /**
+     * Gets the current time.
+     *
+     * @return \DateTimeImmutable
+     *    The current time.
+     * @throws \Tuf\Exception\FormatException
+     *    Thrown if time format is not valid.
+     */
+    private function getCurrentTime(): \DateTimeImmutable
+    {
+        $fakeNow = '2020-08-04T02:58:56Z';
+        $nowDate = $this->metadataTimestampToDateTime($fakeNow);
+        return $nowDate;
+    }
+
+    /**
+     * Determines if the new root metadata has rotated keys for a role.
+     *
+     * @param \Tuf\Metadata\RootMetadata $previousRootData
+     *   The previous root metadata.
+     * @param \Tuf\Metadata\RootMetadata $newRootData
+     *   The new root metadta.
+     * @param string $role
+     *   The role to check for rotated keys.
+     *
+     * @return boolean
+     *   True if the keys for the role have been rotated, otherwise false.
+     */
+    private function hasRotatedKeys(RootMetadata $previousRootData, RootMetadata $newRootData, string $role)
+    {
+        $previousRole = $previousRootData->getRoles()[$role] ?? null;
+        $newRole = $newRootData->getRoles()[$role] ?? null;
+        if (empty($previousRole) && empty($newRole)) {
```

Comment:
When would `previousRole` or `newRole` ever be empty? I'm assuming that `previousRole` is based off the currently trusted root metadata on disk, which you should always have, no?

Reply:
Thanks for pointing this out.

I guess I was thinking your could have a root metadata that didn't have a 'timestamp' or 'snapshot' role specified. But I realize now you can't. I see in `4.3. File formats: root.json` of the spec these are required.

I removed this check and added a validation constraint in `\Tuf\Metadata\RootMetadata::getSignedCollectionOptions()` to ensure this is always true.